### PR TITLE
Don't report anything from GaugeBucketCollector metrics until data is present

### PR DESCRIPTION
**Pull request by anoadragon453**
Fixes https://github.com/matrix-org/synapse/issues/8792.

This PR modifies `GaugeBucketCollector` to only report data once it has been updated, rather than initially reporting a value of 0. Fixes zero values being reported for some metrics on startup until a background job to update the metric's value runs later.

### CODE REVIEW

**Review by richvdh — CHANGES_REQUESTED** [review]
seems reasonable.

> Or if there's a way to register, yet hold off on reporting until we have some useful data instead.

Yes, an alternative approach which I think would be simpler would be to initialise `self._metric` to `None`, and make the yield in `collect` conditional. Then you don't need to worry about delaying registration.

**Comment by erikjohnston** [comment]
I think prometheus might do something unhelpful if you don't report metrics? Something like using the last known value? There's something odd like that, but I can't remember what

**Comment by richvdh** [comment]
> Something like using the last known value?

That would probably be fine here...

File: changelog.d/8926.bugfix [hunk 1]
```diff
@@ -0,0 +1 @@
+Prevent `synapse_forward_extremities` and `synapse_excess_extremity_events` metrics from initially reporting zero-values after startup.
```
Comment (richvdh) [inline changelog.d/8926.bugfix:1]:
```suggestion
Prevent `synapse_forward_extremities` and `synapse_excess_extremity_events` Prometheus metrics from initially reporting zero-values after startup.
```

**Review by richvdh — APPROVED** [review]
lgtm otherwise

**Comment by richvdh** [comment]
actually: can you update the PR summary and description to reflect the updated approach, please?

**Pull request merged**